jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/BPMN2DataServiceImpl.java

```diff
@@ -131,9 +83,29 @@ private void validateNonEmptyDeploymentIdAndProcessIdAndTaskName(String deployme
             throw new IllegalStateException( msg.toString() );
         }
     }
+    
+    public void addProcessDefinition(String deploymentId, String processId, Object processDescriptor, KieContainer kieContainer) {
+        Map<String, ProcessDescriptor> definitions = null;
+        synchronized (definitionCache) {
+            definitions = definitionCache.get(deploymentId);
+            if (definitions == null) {
```

Comment:
putIfAbsent is *meant* for this situation. :) 

1. Only add the map if the map is not there! 
2. Otherwise, use the existing map (that another thread got to first)

Lastly, this is being used when we add a process definition -- given that we want to support scalable and multi-tenant solutions, multiple concurrent deployment requests (I add a process and deploy, you add a different new process and also want to deploy) going to be more and more likely, IMHO.